File: nodecsharp/Services/DirectoryManager.cs
using System.Collections.Concurrent;

namespace CacheNode.Services;

public sealed class DirectoryManager
{
    private readonly object _directoryLock = new();
    private readonly ConcurrentDictionary<string, HashSet<int>> _presenceList = new(StringComparer.Ordinal);
    private readonly ConcurrentDictionary<string, string> _mainMemory = new(StringComparer.Ordinal);
    private readonly ConcurrentDictionary<string, SemaphoreSlim> _variableWriteQueues = new(StringComparer.Ordinal);
    private readonly ILogger<DirectoryManager> _logger;

    public DirectoryManager(ILogger<DirectoryManager> logger)
    {
        _logger = logger;
    }

    public async Task<T> WithVariableWriteLockAsync<T>(string variableName, Func<Task<T>> work)
    {
        var queue = _variableWriteQueues.GetOrAdd(variableName, _ => new SemaphoreSlim(1, 1));
        await queue.WaitAsync();
        try
        {
            return await work();
        }
        finally
        {
            queue.Release();
        }
    }

    public void RegisterVariablePresence(string variableName, int nodeId)
    {
        lock (_directoryLock)
        {
            var owners = _presenceList.GetOrAdd(variableName, _ => new HashSet<int>());
            owners.Add(nodeId);
        }
    }

    public void UpdateMainMemoryValue(string variableName, string value)
    {
        _mainMemory[variableName] = value;
    }

    public IReadOnlySet<int> GetOwnersOf(string variableName)
    {
        lock (_directoryLock)
        {
            return _presenceList.TryGetValue(variableName, out var owners)
                ? owners.ToHashSet()
                : new HashSet<int>();
        }
    }

    public void Clear()
    {
        lock (_directoryLock)
        {
            _presenceList.Clear();
            _mainMemory.Clear();
        }

        _logger.LogInformation("DirectoryManager state cleared before reconstruction.");
    }

    public void RemoveNodeFromPresence(int nodeId)
    {
        lock (_directoryLock)
        {
            foreach (var (_, owners) in _presenceList)
            {
                owners.Remove(nodeId);
            }
        }
    }

    public object Snapshot()
    {
        lock (_directoryLock)
        {
            var presence = _presenceList.ToDictionary(
                item => item.Key,
                item => item.Value.OrderBy(id => id).ToArray(),
                StringComparer.Ordinal);

            var memory = _mainMemory.ToDictionary(item => item.Key, item => item.Value, StringComparer.Ordinal);

            return new
            {
                presenceList = presence,
                mainMemory = memory
            };
        }
    }
}

File: nodepython/app/directory_manager.py
import logging
import threading

logger = logging.getLogger(__name__)


class DirectoryManager:
    def __init__(self):
        self._presence_list = {}
        self._main_memory = {}
        self._lock = threading.RLock()

    def register_variable_presence(self, variable_name: str, node_id: int):
        with self._lock:
            owners = self._presence_list.setdefault(variable_name, set())
            owners.add(node_id)
            logger.debug(
                "Directory Manager: registered '%s' presence on Node %s.",
                variable_name,
                node_id,
            )

    def update_main_memory_value(self, variable_name: str, value: str):
        with self._lock:
            self._main_memory[variable_name] = value
            logger.debug(
                "Directory Manager: updated main memory '%s' = '%s'.",
                variable_name,
                value,
            )

    def get_owners_of(self, variable_name: str):
        with self._lock:
            return set(self._presence_list.get(variable_name, set()))

    def get_value_from_main_memory(self, variable_name: str):
        with self._lock:
            return self._main_memory.get(variable_name)

    def clear(self):
        with self._lock:
            self._presence_list.clear()
            self._main_memory.clear()
            logger.info("Directory Manager: cleared state before reconstruction.")

    def snapshot(self):
        with self._lock:
            return {
                "presenceList": {
                    variable_name: sorted(node_ids)
                    for variable_name, node_ids in self._presence_list.items()
                },
                "mainMemory": dict(self._main_memory),
            }


directory_manager = DirectoryManager()
